### PR TITLE
aio: ensure navigation data is only request once

### PR DESCRIPTION
**Pull request by petebacondarwin**
**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: aio/src/app/navigation/navigation.service.spec.ts [hunk 1]
```diff
@@ -1,14 +1,89 @@
...
+      let views3: NavigationViews;
+      service.navigationViews.subscribe(views => views3 = views);
+
+      expect(views2).toBe(views1);
+      expect(views3).toBe(views1);
+    });
+
+
+    it('should do WHAT(?) if the request fails', () => {
+      console.warn('PENDING: NavigationService navigationViews should do WHAT(?) if the request fails');
```
Comment (gkalpak) [inline aio/src/app/navigation/navigation.service.spec.ts]: It seems you found a way to have your pending tests 😛

File: aio/src/app/navigation/navigation.service.ts [hunk 2]
```diff
@@ -52,8 +53,12 @@ export class NavigationService {
   }
 
   private fetchNavigation(): Observable<NavigationViews> {
+    const subject = new AsyncSubject<NavigationViews>();
     // TODO: logging and error handling
-    return this.http.get(NAVIGATION_PATH).map(res => res.json() as NavigationViews);
+    this.http.get(NAVIGATION_PATH)
+             .map(res => res.json() as NavigationViews)
+             .subscribe(subject);
+    return subject.asObservable();
```
Comment (gkalpak) [inline aio/src/app/navigation/navigation.service.ts]: What is the benefit/difference with returning an `AsyncSubject` (vs the previous implementation)?

**Review by gkalpak — APPROVED** [review]

File: aio/src/app/navigation/navigation.service.ts (hunk 2 above)
Comment (petebacondarwin) [inline aio/src/app/navigation/navigation.service.ts]: The problem with the previous implementation is that calls to `http.get()` set up a cold observable. So each separate subscription of that observable triggers a new request to the server.

In the `AppComponent` we bind to `navigationViews` twice, which caused two requests.

What this implementation does is put a subject in between the client and the http observable.
This `AsyncSubject` will only trigger the original request only once and then multicast the resulting stream to all clients.

Comment (gkalpak) [inline aio/src/app/navigation/navigation.service.ts]: Is it desirable that this triggers the request immediately (even if there are no subscribers)?
We could keep it cold and still share the subscription with something like:

```ts
return this.http.get(...).map(...).publishLast(1).refCount();
```

Comment (DzmitryShylovich) [inline aio/src/app/navigation/navigation.service.ts]: `.share()`?

Comment (gkalpak) [inline aio/src/app/navigation/navigation.service.ts]: That too (since `.get()` will only return one value) :smiley:

Comment (petebacondarwin) [inline aio/src/app/navigation/navigation.service.ts]: trouble with `refCount` and friends is that if all the subscribers unsubscribe then the whole observable will be unsubscribed. Then if someone reconnects the original request will be made again.
So we have to make a decision... we could make our own version of `refCount` that connected when the subscribers went from 0 to 1 and then did not disconnect when the subscribers went from 1 to 0. But is it really worth it?
I think that we would want to have this data ASAP in the real app.

**Comment by petebacondarwin** [comment]
Closing in favour of #14940 which supersedes it

**Pull request closed without merging**